### PR TITLE
added first cc computation throw

### PR DESCRIPTION
**Pull request by MaxOstrowski**
Do not merge, just to collect things:
- [x]  still crashes on full propagation on 2sp BENG01
- [x]  currently relies on order of vertices in set, possible but not nice
- [x]  need to test multi shot cc computation
- [ ]  changed search deterministic (different number of choices/conflicts vs wip branch), should not happen or find a reason why
- [x] change outgoing to new nodes, otherwise Mutexes could be broken

### CODE REVIEW

**Comment by rkaminsk** [comment]
You might have noticed that the unit tests fail.

**Comment by MaxOstrowski** [comment]
@rkaminsk I have a question regarding determinism.
My algorithm changes the in/out node of edges in `edges_` from node "0" to other nodes "null(x)".
I can't see why this changes determinism, i.e. number of choices/conflicts. Any idea? Have I done something wrong ?

**Comment by rkaminsk** [comment]
There might be multiple cycles when adding a new edge. If you change the order in which edges are traversed, then you might get a different conflicting clause. Also, in the complete graph the algorithm might change potentials earlier. Maybe this can lead to different traversals of the graph, too. I don't know if the last point really can lead to different conflicting clauses. This would require some thinking...

**Comment by MaxOstrowski** [comment]
> There might be multiple cycles when adding a new edge.

I do not get this point. I'm not adding edges, I simply replace a node index with another one. No cycles should be introduced.

> If you change the order in which edges are traversed, then you might get a different conflicting clause. 

That's my presumption. But I do not know what could have caused this change, as all edges stay exactly the same. In Propagation and check the cc part just removes dead ends that never lead to a cycle or any other implication. I will have to check more thoroughly then...

**Comment by rkaminsk** [comment]
I also updated the CI tests to github actions. Best merge this in.

**Comment by MaxOstrowski** [comment]
Here some preliminary results:
[results.zip](https://github.com/potassco/clingoDL/files/4522881/results.zip)


- [ ]  Maybe the benchmark settings are still fragile ? In the default configuration the scc and the wip branch have a big difference in instance flow_shop/tai50_20_3 although dl propagation time seems farily equal and there is just one cc. Same amount of choices/constraints. This difference is not explainable.
- [ ] Initialization time for computing cc's seems to be negligible
- [ ] No difference in choices/constraints if propagation is disabled
- [ ] Checking constraints seems to be speed up (without propagation) see benchmark BENG07. Otherwise are there rarely any benchmarks where the constraint checking takes time
- [ ] Comparing with configuration --propagate=full on examples with cc=1 seems also unclear. Examples flowshow - tai 50_20_3 and 50_10_3. Search seems to be exactly equal, even added and skipped edges in dl. Nevertheless, csp propagation times seem to vary (of by ~5secs) and total time is off by ~40s. How? Benchmark Environment: `<pbsjob name="pbs-gen" timeout="900" runs="1" script_mode="timeout" walltime="0:59:59" cpt="4"/>`
- [ ] Propagation speed on examples with cc>1 seems to be improved

**Comment by rkaminsk** [comment]
> Comparing with configuration --propagate=full on examples with cc=1 seems also unclear. Examples flowshow - tai 50_20_3 and 50_10_3. Search seems to be exactly equal, even added and skipped edges in dl. Nevertheless, csp propagation times seem to vary (of by ~5secs) and total time is off by ~40s. How? Benchmark Environment: `<pbsjob name="pbs-gen" timeout="900" runs="1" script_mode="timeout" walltime="0:59:59" cpt="4"/>`

I do not know where you are running the benchmark. If you run on our cluster, then you should consider that the machines in the default partition have only 2 numa nodes. To get clean results, you can set cpt=24. Then nodes in the kr partition just have 1 numa node. Set cpt=8 for clean results.

**Comment by MaxOstrowski** [comment]
[results.zip](https://github.com/potassco/clingoDL/files/4527441/results.zip)

Redone benchmarks with cpt=8 on krnodes (zuse).
Now it is "instances/flow_shop/tai100_20_3.lp" which has 100s difference between both branches.
All stats are equal except time. This renders the benchmark useless. Will investigate.

**Comment by MaxOstrowski** [comment]
[watcher.zip](https://github.com/potassco/clingoDL/files/4527487/watcher.zip)

I looks like another process was running on the same node ? As the loadavg is 2 ?
This, and less free memory are the only hints I could find.

**Comment by MaxOstrowski** [comment]
[results.zip](https://github.com/potassco/clingoDL/files/4529663/results.zip)
And again. This time deviation is much smaller. Some heuristic effects, no measureable speedup for instances with more than 1 cc. But it does not seem to be slower and may sometimes help.

**Comment by rkaminsk** [comment]
> [results.zip](https://github.com/potassco/clingoDL/files/4529663/results.zip)
> And again. This time deviation is much smaller. Some heuristic effects, no measureable speedup for instances with more than 1 cc. But it does not seem to be slower and may sometimes help.

I don't know what you mean by it is slower but sometimes help. I also did not find the number of CCs in the table. So I do not know which instances are affected.

**Comment by MaxOstrowski** [comment]
clingodlcc_full.ods you see the number of ccs in the configurations ran with the 267a7d4 branch (scc branch) in the column "ccs".
Some of the train instances and all 2sp instances.
I meant that it is *not* slower (csptimeinit does not go up to split the graph into several chunks) on the instances that only have one cc. So I would recommend that this feature is not an command line option and always on.
Any complaints about the code ? Otherwise I would merge to wip.

**Comment by rkaminsk** [comment]
> clingodlcc_full.ods you see the number of ccs in the configurations ran with the [267a7d4](https://github.com/potassco/clingoDL/commit/267a7d489dbb2e29a6baaeb192aad07ef9e867a0) branch (scc branch) in the column "ccs".
> Some of the train instances and all 2sp instances.
> I meant that it is _not_ slower (csptimeinit does not go up to split the graph into several chunks) on the instances that only have one cc. So I would recommend that this feature is not an command line option and always on.
> Any complaints about the code ? Otherwise I would merge to wip.

Let me have a look at the code first. I already spotted some strange things like:

```cpp
struct CC {
    CC(int cc, bool used) : cc(cc), used(used) {}
    size_t cc : 31;
    size_t used : 1;
};
```

You should use `uint32_t` here. It also looks like there is a mess between signed and unsigned types here. I have not looked in detail but it also looks like the zero node would better be stored in the CC struct then in a set, right?

**Comment by MaxOstrowski** [comment]
I agree with the signed types.
Storing an extra flag in CC for zero nodes would be possible. But there are still places where we need to iterate the zero nodes, and therefore need some "central storage" or some kind of "collection". We could put the zero_node flag to CC and replace the set with a vector.
Or we completely remove the set and use the clingo symbols, these can always be recreated as we know how to iteratively generate their names. But I do not know if this version is as clean...

**Comment by rkaminsk** [comment]
Have you considered the multi-shot case? I think it will fail if there is an edge in the multi-shot case that joins two components. Then null nodes would have to be joined again.

**Comment by MaxOstrowski** [comment]
This should work and I have a small test case for it, but no larger checks. As combining CC's can be linear, i simply recompute the cc's.

**Comment by rkaminsk** [comment]
The way you handle incoming edges still looks fishy. For example you clear the incoming edges but then do not completely recalculate them. Maybe we should forgo any optimizations here and compute these vectors on every step not even having them as global variables. Just keeping the edge_start offset for the mutex calculation is important.

**Comment by MaxOstrowski** [comment]
Removing incoming was wrong, is now fixed. Unsure about removing optimizations, but it will save some space during solving. outgoing and incoming edges are local variables now.

**Comment by rkaminsk** [comment]
Some more things that should be addressed:
- Please put opening parentheses on the same line as the function.
- I put `const` after the type and `*` or `&` on the name. You can either match my style or, since I don't care that much about this, change my code to match your style. What I do care about is consistency.
- Please put a space after while.
- Maybe `node_info_`/`NodeInfo` is a better name than `cc_`/`CC`.
- I think that `valid_cc_non_zero_node` is a horrible name. Maybe `cc_visited`? If you want to describe more just put a comment.
- Please put spaces around binary operators.
- The set `zero_nodes_` must be a vector to get constant lookups. Even though it is unlikely to matter much in practice, you made lower bound lookups linear: `auto zero_node = *(std::next(zero_nodes_.begin(),cc+1))`. All you have to do is store the component offset for zero nodes in the `cc_` vector, too. The test whether something is a null node becomes `zero_nodes_[cc_[idx].cc] == idx`.

**Comment by MaxOstrowski** [comment]
Had to change more than I intended,
also readded the outgoing adjacency fix if it is reused for the mutexes. Should not cost much.
I also do it for incoming, although its no longer needed.

**Comment by rkaminsk** [comment]
Please reorder the members in the class so that there are no holes because of alignment. In C++ we unfortunately not always group members logically.

On a second note I find it really strange that you are mixing flag based and tree based sets. Performance dose not matter much for this part of the algorithm, so you can choose either way. Just do not mix.

For a depth first search one flag is enough. Only enqueue something if the visited flag is not set and mark it right away as visited. Then you know that anything in the queue was not yet visited and you also won't enqueue unnecessary nodes. It is really important to get algorithms down to the essential bits and avoid all the unnecessary cruft. Not just for performance but also for readability.

**Comment by MaxOstrowski** [comment]
I hope I got everything right.
- I reordered the member in the DLPropagator class, it is now smaller, but I do not know if it is already perfect.
- I removed the set for checking the visited nodes and marked them directly. I hope this is what you meant with not mixing. Otherwise please give some more pointers.

Thanks a lot.

**Comment by rkaminsk** [comment]
My approach regarding alignment is to simply put the 64bit aligned stuff first. This includes STL containers and most classes, then follow with 32 bit types, and then the 8bit types. This way all holes will be at the end, which makes extending the class easier because one does not have to look where to insert members.

Instead of `while (node_stack.size())` I would always write `while (!node_stack.empty())`. Linters will also warn about this and even some compilers. It is easier to read.

The maps
```cpp
std::unordered_multimap<int, int> outgoing_change; 
std::unordered_multimap<int, int> incoming_change; 
```
can simply be vectors of pairs.

I think that since you call `node_info_.clear()` in the beginning you have to reinsert the zero nodes afterward.

**Comment by MaxOstrowski** [comment]
- redid the alignment
- used empty, (looked it up, nice to know)
- replaced maps with vector
- readded zero_nodes

> PS: Don't despair! We are almost there!

Don't worry, just be patient with me :smiling_face_with_three_hearts:

**Comment by rkaminsk** [comment]
The algorithm looks very readable now. One can immediately see that it is just a depth-first search. :wink: 

One last thing we should remove is the zero_node member (which as a member should also end with _ but probably this was my slight because I did not pay too much attention when hacking on clingo-dl). Anyway, it is better to remove it. Just put it initially in the `zero_nodes_` vector and ensure that it is in `node_info_`. All this can happen in the constructor. Then you can simply start the loop in [Line 1288](https://github.com/potassco/clingoDL/blob/2aeb850a4316726df528cadbbec1c398949c8399/libclingo-dl/clingo-dl/propagator.hh#L1228) with one. This eliminates all the annoying special cases. Don't forget to remove the pop_back.

**Comment by MaxOstrowski** [comment]
> One can immediately see that it is just a depth-first search.

:facepalm: 
> This eliminates all the annoying special cases.

I'm not sure about it, i deliberately left it out so the cc corresponds to the array index.
The node map_vert(Clingo::Number(0)) has no cc it belongs to, and would therefore introduce a special case itself, no ?

File: libclingo-dl/clingo-dl/propagator.hh [hunk 1]
```diff
@@ -1158,6 +1174,90 @@ public:
         }
     }
 
+    bool cc_visited(int node) const {
+        return node_info_[node].visited == cc_visited_;
+    }
+
+    bool is_zero(int node) const {
+        return node == zero_node || (node_info_[node].cc < zero_nodes_.size() && zero_nodes_[node_info_[node].cc] == node);
```
Comment (rkaminsk) [inline libclingo-dl/clingo-dl/propagator.hh:1182]: Here is an unnecessary special case.

File: libclingo-dl/clingo-dl/propagator.hh [hunk 2]
```diff
@@ -1158,6 +1174,90 @@ public:
...
+        for (int i = 0; i < zero_nodes_.size(); ++i) {
+            auto node = map_vert(Clingo::Function("__null", {Clingo::Number(i)}));
+            zero_nodes_[i] = node;
+            node_info_.resize(std::max(node_info_.size(), static_cast<size_t>(node+1)), NodeInfo(0, !cc_visited_));
+            node_info_[node] = NodeInfo(i, cc_visited_);
+        }
+
+        std::vector< std::pair<int, int> > outgoing_change; 
+        std::vector< std::pair<int, int> > incoming_change; 
+        zero_nodes_.emplace_back(map_vert(Clingo::Number(0)));
```
Comment (rkaminsk) [inline libclingo-dl/clingo-dl/propagator.hh:1237]: here is an unnecessary special case

File: libclingo-dl/clingo-dl/propagator.hh [hunk 3]
```diff
@@ -1158,6 +1174,90 @@ public:
...
+                auto& e = edges_[edge->second];
+                auto cc = node_info_[e.from].cc;
+                e.to = zero_nodes_[cc];
+                incoming_change.emplace_back(zero_nodes_[cc], edge->second);
+            }
+            incoming.erase(range.first, range.second);
+        }
+        outgoing.insert(outgoing_change.begin(), outgoing_change.end());
+        incoming.insert(incoming_change.begin(), incoming_change.end());
+        zero_nodes_.pop_back();
```
Comment (rkaminsk) [inline libclingo-dl/clingo-dl/propagator.hh:1258]: here is an unnecessary special case

File: libclingo-dl/clingo-dl/propagator.hh [hunk 4]
```diff
@@ -1370,11 +1477,14 @@ private:
...
     std::unordered_multimap<literal_t, int> false_lit_to_edges_;
     std::vector<Edge<T>> edges_;
-    std::unordered_multimap<int, int> outgoing_;
     std::vector<Clingo::Symbol> vert_map_;
     std::unordered_map<Clingo::Symbol, int> vert_map_inv_;
+    std::vector<NodeInfo> node_info_;
+    std::vector<int> zero_nodes_; // all nodes that are equivalent to 0 except zero_node Clingo::Number(0)
     Stats &stats_;
     PropagatorConfig conf_;
+    int zero_node;
```
Comment (rkaminsk) [inline libclingo-dl/clingo-dl/propagator.hh:1486]: this value is unnecessary

**Comment by rkaminsk** [comment]
> > One can immediately see that it is just a depth-first search.
> 
> facepalm
> 
> > This eliminates all the annoying special cases.
> 
> I'm not sure about it, i deliberately left it out so the cc corresponds to the array index.
> The node map_vert(Clingo::Number(0)) has no cc it belongs to, and would therefore introduce a special case itself, no ?

It has another name than the other ones. But his is the *only* difference. I already highlighted the line above where it is easy to handle the name difference. Since we have an array of zero nodes we should put this value in there too.

**Comment by MaxOstrowski** [comment]
I know reserved the special cc=0 which is always empty except for Clingo::Number(0).

Im running the test_framework and one instance of 2SP with valgrind as checks ...
Still, I'm not sure if things got any better now :D

**Comment by rkaminsk** [comment]
Am I missing something? Why do you want to have an empty CC? Simply start with zero. Just the loop in line 1228 has to start from 1. With this if there is just one component, you don't even have to introduce any new nodes and variable `0` does not hang around uselessly.

**Comment by MaxOstrowski** [comment]
My problem is the is_zero check. There is use node_info to obtain a cc for this node.
As Clingo::Number(0) is a zero node it has to have a cc, no ? I'm puzzled.

**Comment by rkaminsk** [comment]
> My problem is the is_zero check. There is use node_info to obtain a cc for this node.
> As Clingo::Number(0) is a zero node it has to have a cc, no ? I'm puzzled.

But it will just have cc 0. You don't even have to do anything. This happens automatically.

**Comment by MaxOstrowski** [comment]
I'm just a bit slow in the mind, think I now got it.

**Comment by rkaminsk** [comment]
There are still some unnecessary tests like `!state.dl_graph.empty() && state.dl_graph.valid_node(node)` but I will take it from here. I'll also make a pass over the code and then merge. It might take a little. I can prioritize if it is blocking you. In any case, a clingo-dl release should probably wait until the next clingo release. This way we can announce everything together. Thanks for the work.

**Comment by MaxOstrowski** [comment]
Thanks a lot.
This is not blocking me, so take your time and do it when you like, but preferably before the release.

Thanks a lot for guiding me through :+1:

**Comment by rkaminsk** [comment]
I still found some small things to refactor. Please tell me if this is okay and I'll merge.

**Comment by MaxOstrowski** [comment]
(Only) Sometimes I feel stupid.
Cool that you managed to removed cc_visited.
Looks okay, free to merge.

**Comment by rkaminsk** [comment]
I don't think it has anything to do with being stupid. It's more about taking a deep breath, concentrate, and then only write down what is necessary. :wink:

**Comment by rkaminsk** [comment]
I accidentally wrote strongly connected components. Read it as connected components.

**Pull request merged**